### ISSUE TITLE
BUG: np.union1d(ar1, ar2) fails to flatten its arguments (as promised in its docstring)

### ISSUE THREAD

**Issue by WarrenWeckesser** [numpy/numpy#10340]
This was discovered while looking into [a question on stackoverflow](https://stackoverflow.com/questions/48139632/python-setdiff1d-is-wotking-but-union1d-is-not-working-for-the-same-data).  According to the docstring, the arguments to `numpy.union1d(ar1, ar2)` "are flattened if they are not already 1D."  But in fact, they are not, and an error is raised if the shapes of the arguments are not compatible for concatenation:

```
In [40]: x
Out[40]: 
array([[0, 1, 2],
       [3, 4, 5]])

In [41]: y
Out[41]: array([0, 1, 2, 3, 4])

In [42]: np.union1d(x, y)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-42-14b2c2e5e9db> in <module>()
----> 1 np.union1d(x, y)

~/miniconda3scipy/lib/python3.5/site-packages/numpy/lib/arraysetops.py in union1d(ar1, ar2)
    612     array([1, 2, 3, 4, 6])
    613     """
--> 614     return unique(np.concatenate((ar1, ar2)))
    615 
    616 def setdiff1d(ar1, ar2, assume_unique=False):

ValueError: all the input arrays must have same number of dimensions
```

The complete definition of `numpy.union1d` is

```
def union1d(ar1, ar2):
    return unique(np.concatenate((ar1, ar2))) 
```

so this bug could probably be tagged as an easy fix.

**Comment by eric-wieser** [comment]
Perhaps it should become
```
def union(arr1, arr2, axis=None):
    return unique(np.concatenate((ar1, ar2), axis=axis), axis=axis)
```
(note the rename)

**Comment by eric-wieser** [comment]
Documentation is now fixed as of #10342. Do we still want to add an `axis` argument to `union`?
